### PR TITLE
refactor: resolve promises

### PR DESCRIPTION
**Pull request by TriForMine**

### CODE REVIEW

File: src/rest/request.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export function createRequestBody(queuedRequest: QueuedRequest) {
   // IF A REASON IS PROVIDED ENCODE IT IN HEADERS
   if (queuedRequest.payload.body?.reason) {
     headers["X-Audit-Log-Reason"] = encodeURIComponent(
-      queuedRequest.payload.body.reason,
+      queuedRequest.payload.body?.reason,
```
Comment (ayntee) [inline src/rest/request.ts:55]: This is unnecessary change.

**Comment by ayntee** [comment]
Can you spin `deno lint src/** test/** --unstable --ignore=./src/types/`?

File: src/rest/request.ts (hunk 1 above)
Comment (TriForMine) [reply src/rest/request.ts]: Reverted

File: src/ws/shard.ts [hunk 2]
```diff
@@ -15,8 +15,9 @@ import {
 import { BotStatusRequest, delay } from "../util/utils.ts";
 import { decompressWith } from "./deps.ts";
 import { handleDiscordPayload } from "./shard_manager.ts";
+import { Collection } from "../util/collection.ts";
 
-const basicShards = new Map<number, BasicShard>();
+const basicShards = new Collection<number, BasicShard>();
```
Comment (ayntee) [inline src/ws/shard.ts:20]: New Collection, why? We don't really need extra methods

Comment (TriForMine) [reply src/ws/shard.ts:20]: It's needed for Promise.all at this line: https://github.com/TriForMine/discordeno/blob/70a3eef41aac9407c33e940584f584de8d6f605f/src/ws/shard.ts#L331
I'm not sure if there is an other way to do it.

File: src/ws/shard.ts [hunk 3]
```diff
@@ -327,7 +328,7 @@ async function processGatewayQueue() {
     return;
   }
 
-  basicShards.forEach((shard) => {
+  await Promise.all(basicShards.map(async (shard) => {
```
Comment (ayntee) [inline src/ws/shard.ts:331]: This is unnecessary, since we are not really returning a promise from the .map()

Comment (TriForMine) [reply src/ws/shard.ts:331]: It's actually needed since to handle the rate limit we need to wait until everything is done and then wait 2 seconds if I'm not wrong.

File: src/ws/shard_manager.ts [hunk 4]
```diff
@@ -40,7 +40,7 @@ export async function spawnShards(
       data.shards > lastShardID ? data.shards : lastShardID,
     ];
     // Start The shard
-    await createShard(data, payload, false, shardID);
+    createShard(data, payload, false, shardID);
```
Comment (ayntee) [inline src/ws/shard_manager.ts:43]: `await` removed, can you explain?

Comment (TriForMine) [reply src/ws/shard_manager.ts:43]: createShard isn't an async function 🤔

File: src/ws/shard_manager.ts [hunk 5]
```diff
@@ -99,7 +99,15 @@ export function requestAllMembers(
...
-  return requestGuildMembers(guild.id, guild.shardID, nonce, options);
+
+  const result = await requestGuildMembers(
+    guild.id,
+    guild.shardID,
+    nonce,
+    options,
+  );
+
+  return result;
```
Comment (ayntee) [inline src/ws/shard_manager.ts:110]:
```suggestion
  await requestGuildMembers(
    guild.id,
    guild.shardID,
    nonce,
    options,
  );
```
Not returning anything also returns undefined, so basically the assignment is redundant.

**Review by ayntee — APPROVED** [review]
LGTM

File: test/deps.ts [hunk 6]
```diff
@@ -2,5 +2,5 @@ export {
   assertArrayIncludes,
   assertEquals,
   assertExists,
-} from "https://deno.land/std@0.85.0/testing/asserts.ts";
+} from "https://deno.land/std@0.86.0/testing/asserts.ts";
```
Comment (ayntee) [inline test/deps.ts:5]: Can you remove this change?

**Review by TriForMine — COMMENTED** [review]
_ignore this_

**Review by ayntee — APPROVED** [review]
LGTM!
